### PR TITLE
Avoid applying the protobuf plugin before its dependencies

### PR DESCRIPTION
**Pull request by steffenyount**
fixes google/protobuf-gradle-plugin#40

The gist of this patch is that the protobuf plugin will now use the
PluginManager.withPlugin() callback mechanism to delay applying
itself until at least one of its prerequisite plugins has been
applied.

I was able to reproduce the issue and verify that this patch fixes it
when used with Gradle 2.4 and the java plugin.

I also verified that this patch overcomes the new plugins {} DSL's
early apply phase by using my newly published 'plugin-delegator'
plugin test harness which itself gets called during that same early
apply phase to then delegates apply() calls to whatever pluginIds
(local, custom, or otherwise) were specified for it.

### CODE REVIEW

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 1]
```diff
@@ -53,42 +62,65 @@ class ProtobufPlugin implements Plugin<Project> {
...
-          linkGenerateProtoTasksToJavaCompile()
-          // protoc and codegen plugin configuration may change through the protobuf{}
-          // block. Only at this point the configuration has been finalized.
-          project.protobuf.tools.resolve()
-        }
+      addSourceSetExtensions()
+      getSourceSets().all { sourceSet ->
+        createConfiguration(sourceSet.name)
+      }
+
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: The following content should stay in `project.afterEvaluate`. Otherwise they will be run as soon as the prerequisite applies, at which point the `protobuf` DSL may have not been evaluated yet.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 2]
```diff
@@ -53,42 +62,65 @@ class ProtobufPlugin implements Plugin<Project> {
...
     void apply(final Project project) {
-        this.project = project
-        def gv = project.gradle.gradleVersion =~ "(\\d*)\\.(\\d*).*"
-        if (!gv || !gv.matches() || gv.group(1).toInteger() < 2 || gv.group(2).toInteger() < 2) {
-            println("You are using Gradle ${project.gradle.gradleVersion}: This version of the protobuf plugin requires minimum Gradle version 2.2")
+      this.project = project
+      def gv = project.gradle.gradleVersion =~ "(\\d*)\\.(\\d*).*"
+      if (!gv || !gv.matches() || gv.group(1).toInteger() < 2 || gv.group(2).toInteger() < 2) {
+          println("You are using Gradle ${project.gradle.gradleVersion}: This version of the protobuf plugin requires minimum Gradle version 2.2")
+      }
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: I know the indentation is inconsistent everywhere, but please keep the format as is and let the commit contain only the meaty content. Formatting should be addressed as a separate commit.

Comment (steffenyount) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: Ok no problem, I will revert that part of my change.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy (hunk 1 above)
Comment (steffenyount) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: I will revert this part as well.

**Comment by zhangkun83** [comment]
I have published a `0.7.4.-testing1" version to [plugins.gradle.org](https://plugins.gradle.org/plugin/com.google.protobuf), and this seems to work on Gradle 2.4+.

It would fail with the following error with Gradle 2.3 when trying to apply the Java plugin:
```
An exception occurred applying plugin request [id: 'java']
> Failed to apply plugin [id 'org.gradle.java']
   > Could not find property 'sourceSets' on root project 'sample-gradle-project'
```

Could you verify if it's also the case for you? Below is how I applied the plugins:
```
plugins {
    id "java"
    id "com.google.protobuf" version "0.7.4-testing1"
}
```

If you can verify it, we should raise the Gradle version requirement to 2.4

**Comment by steffenyount** [comment]
yes, I have verified. I do get the same results. It fails on 2.4 with the same error as you reported for the existing version but it passes with no such error using the '0.7.4-testing1' version.

**Comment by zhangkun83** [comment]
> It fails on 2.4 with the same error as you reported for the existing version

I am confused. Do you mean with 2.4 and 0.7.3 it would fail with the `Could not find property 'sourceSets' on root project 'sample-gradle-project'` error? That was not I expected.

**Comment by steffenyount** [comment]
Let me try to restate what I said:

0.7.3 fails with

```
* What went wrong:
An exception occurred applying plugin request [id: 'com.google.protobuf', version: '0.7.3']
> Failed to apply plugin [id 'com.google.protobuf']
   > Please apply the Java plugin or the Android plugin first. This error may also be a result of using the new Gradle plugins DSL. Please use the traditional "apply" function as described in README.md.
```
and 0.7.4-testing1 passes

**Comment by zhangkun83** [comment]
Understood. What you have seen is expected, while I was talking about a different case, where I ran the plugin under Gradle 2.3 and got `Could not find property 'sourceSets' on root project`. If you try with Gradle 2.3 and get the same error, we should raise the Gradle version requirement of the plugin, which is currently [2.2](https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L58), to 2.4.

**Comment by steffenyount** [comment]
I think setting the minimum version requirement to 2.4 is a good idea.

Here are test results from building a simple 'sample-gradle-project' consisting of only the gradle wrapper and a single root 'build.gradle' file with the following contents:
```
plugins {
    id "java"
    id "com.google.protobuf" version "0.7.4-testing1"
}
```

I ran the following bash command on OSX to test:
```
for ((ii=3; ii<=10; ii++)); do
  echo;
  echo "*** Test using Gradle 2.${ii} ***";
  echo;
  sed -i '' -e 's#gradle-2.[0-9]*-#gradle-'"2.${ii}"'-#' gradle/wrapper/gradle-wrapper.properties;
  ./gradlew clean jar;
done
```

and I got the following output:
```
*** Test using Gradle 2.3 ***


FAILURE: Build failed with an exception.

* Where:
Build file '/Users/steffenyount/proj/sample-gradle-project/build.gradle' line: 2

* What went wrong:
An exception occurred applying plugin request [id: 'java']
> Failed to apply plugin [id 'org.gradle.java']
   > Could not find property 'sourceSets' on root project 'sample-gradle-project'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 2.737 secs

*** Test using Gradle 2.4 ***

:clean
:extractIncludeProto UP-TO-DATE
:extractProto UP-TO-DATE
:generateProto UP-TO-DATE
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar

BUILD SUCCESSFUL

Total time: 3.699 secs

This build could be faster, please consider using the Gradle Daemon: http://gradle.org/docs/2.4/userguide/gradle_daemon.html

*** Test using Gradle 2.5 ***

:clean
:extractIncludeProto UP-TO-DATE
:extractProto UP-TO-DATE
:generateProto UP-TO-DATE
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar

BUILD SUCCESSFUL

Total time: 3.719 secs

This build could be faster, please consider using the Gradle Daemon: http://gradle.org/docs/2.5/userguide/gradle_daemon.html

*** Test using Gradle 2.6 ***

:clean
:extractIncludeProto UP-TO-DATE
:extractProto UP-TO-DATE
:generateProto UP-TO-DATE
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar

BUILD SUCCESSFUL

Total time: 3.723 secs

This build could be faster, please consider using the Gradle Daemon: https://docs.gradle.org/2.6/userguide/gradle_daemon.html

*** Test using Gradle 2.7 ***

:clean
:extractIncludeProto UP-TO-DATE
:extractProto UP-TO-DATE
:generateProto UP-TO-DATE
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar

BUILD SUCCESSFUL

Total time: 3.638 secs

This build could be faster, please consider using the Gradle Daemon: https://docs.gradle.org/2.7/userguide/gradle_daemon.html

*** Test using Gradle 2.8 ***

:clean
:extractIncludeProto UP-TO-DATE
:extractProto UP-TO-DATE
:generateProto UP-TO-DATE
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar

BUILD SUCCESSFUL

Total time: 3.518 secs

This build could be faster, please consider using the Gradle Daemon: https://docs.gradle.org/2.8/userguide/gradle_daemon.html

*** Test using Gradle 2.9 ***

:clean
:extractIncludeProto UP-TO-DATE
:extractProto UP-TO-DATE
:generateProto UP-TO-DATE
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar

BUILD SUCCESSFUL

Total time: 3.544 secs

This build could be faster, please consider using the Gradle Daemon: https://docs.gradle.org/2.9/userguide/gradle_daemon.html

*** Test using Gradle 2.10 ***

:clean
:extractIncludeProto UP-TO-DATE
:extractProto UP-TO-DATE
:generateProto UP-TO-DATE
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar

BUILD SUCCESSFUL

Total time: 3.896 secs

This build could be faster, please consider using the Gradle Daemon: https://docs.gradle.org/2.10/userguide/gradle_daemon.html
```

**Comment by zhangkun83** [comment]
Thanks for the exhaustive testing! Please update the version checking code in `ProtobufPlugin.groovy`.

**Comment by steffenyount** [comment]
the version checking code has been updated

**Comment by zhangkun83** [comment]
Can you please sign the [CLA](https://cla.developers.google.com/clas)? Otherwise I won't be able to accept your code.

**Comment by steffenyount** [comment]
done. :-)

**Comment by zhangkun83** [comment]
Great. Thanks!
